src/v1_0/rws_client.cpp

```diff
@@ -432,51 +433,59 @@ std::string RWSClient::getResourceURI(OperationModeResource const&) const
 
 void RWSClient::processEvent(Poco::AutoPtr<Poco::XML::Document> doc, SubscriptionCallback& callback) const
 {
-  // IMPORTANT: don't use AutoPtr<XML::Node> here! Otherwise you will get memory corruption.
-  Poco::XML::Node const * li_node = doc->getNodeByPath("html/body/div/ul/li");
-  if (!li_node)
-    BOOST_THROW_EXCEPTION(ProtocolError {"Cannot parse RWS event message: can't find XML path html/body/div/ul/li"});
+  // IMPORTANT: don't use AutoPtr<XML::Element> here! Otherwise you will get memory corruption.
+  Poco::XML::Element const * ul_element = dynamic_cast<Poco::XML::Element const *>(doc->getNodeByPath("html/body/div/ul"));
+  if (!ul_element)
+    BOOST_THROW_EXCEPTION(ProtocolError {"Cannot parse RWS event message: can't find XML element at path html/body/div/ul"});
 
-  auto const * a_node = li_node->getNodeByPath("a");
-  if (!a_node)
-    BOOST_THROW_EXCEPTION(ProtocolError {"Cannot parse RWS event message: can't find XML path html/body/div/ul/li/a"});
-
-  std::string uri;
-  uri = xmlNodeGetAttributeValue(a_node, "href");
-
-  std::string const class_attribute_value = xmlNodeGetAttributeValue(li_node, "class");
-
-  if (class_attribute_value == "ios-signalstate-ev")
-  {
-    IOSignalStateEvent event;
-    std::string const prefix = "/rw/iosystem/signals/";
-
-    if (uri.find(prefix) != 0)
-      BOOST_THROW_EXCEPTION(ProtocolError {"Cannot parse RWS event message: invalid resource URI"} << UriErrorInfo {uri});
-
-    event.signal = uri.substr(prefix.length(), uri.find(";") - prefix.length());
-    event.value = xmlFindTextContent(li_node, XMLAttribute {"class", "lvalue"});
-    callback.processEvent(event);
-  }
-  else if (class_attribute_value == "rap-ctrlexecstate-ev")
-  {
-    RAPIDExecutionStateEvent event;
-    event.state = rw::makeRAPIDExecutionState(xmlFindTextContent(li_node, XMLAttribute {"class", "ctrlexecstate"}));
-    callback.processEvent(event);
-  }
-  else if (class_attribute_value == "pnl-ctrlstate-ev")
-  {
-    ControllerStateEvent event;
-    event.state = rw::makeControllerState(xmlFindTextContent(li_node, XMLAttribute {"class", "ctrlstate"}));
-    callback.processEvent(event);
-  }
-  else if (class_attribute_value == "pnl-opmode-ev")
+  Poco::AutoPtr<Poco::XML::NodeList> li_elements = ul_element->getElementsByTagName("li");
```

Comment:
I am definitely a newbie here but: Isn't there a risk to have memory corruption here with this AutoPtr ?

Reply:
@MathuxNY-73 the Poco documentation says:

"The returned [NodeList](https://docs.pocoproject.org/current/Poco.XML.NodeList.html) must be released with a call to release() when no longer needed."

https://docs.pocoproject.org/current/Poco.XML.Element.html#32475

Reply:
Actually, my question was more related to the warning at the top of the function that says that we should not use AutoPtr because we will get memory corruption.
I am missing something I am sure but I just would like to understand why it is not dangerous to use it here ?
When checking the documentation this instruction of calling release, indeed, does not exist on the return type of getNodeByPath. Could it mean that this particular method returns a pointer to a part of the document ?
I guess it does

Reply:
This is what it does, yes.